### PR TITLE
Add Z-function algorithm implementation

### PR DESCRIPTION
**Pull request by nikalosa**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: strings/z_function.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+"""
+https://cp-algorithms.com/string/z-function.html#:~:text=The%20Z%2Dfunction%20for%20this,Note.
```
Comment (cclauss) [inline strings/z_function.py:2]:
```suggestion
https://cp-algorithms.com/string/z-function.html
```

**Comment by cclauss** [comment]
One Python file per pull request please.

File: strings/z_function.py (hunk 1 above)
Comment (nikalosa) [reply strings/z_function.py]: Thanks for the review, I will change it

**Comment by nikalosa** [comment]
> One Python file per pull request please.

Sure, it was mistake. I'm going to make another PR for that file and accidentally I pushed here.

File: strings/z_function.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+
+For the value of the first element the algorithm always returns 0
+
+"""
+
+
+def z_function(input_str: str) -> list:
+    """
+    Will convert the entire string to uppercase letters
```
Comment (cclauss) [inline strings/z_function.py:22]: This comment does not describe what this function does.

File: strings/z_function.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+        # if new index's result gives us more right interval, we've to update left_pointer and right_pointer
+        if i + z_result[i] - 1 > right_pointer:
+            left_pointer, right_pointer = i, i + z_result[i] - 1
+
+    return z_result
+
+
+# helping function which checks if following elements are equal or not
+def go_next(i, z_result, s):
```
Comment (cclauss) [inline strings/z_function.py:55]: Can we convert the comment to be the function's docstring.

> checks if following the elements are equal (to what?!?)

File: strings/z_function.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    for i in range(1, len(input_str)):
+        # case when current index is inside the interval
+        if i <= right_pointer:
+            min_edge = min(right_pointer - i + 1, z_result[i - left_pointer])
+            z_result[i] = min_edge
+
+        while go_next(i, z_result, input_str):
+            z_result[i] += 1
+
+        # if new index's result gives us more right interval, we've to update left_pointer and right_pointer
```
Comment (cclauss) [inline strings/z_function.py:47]: Please wrap lines at 88 characters max per line.

**Comment by nikalosa** [comment]
> Is this related to https://en.wikipedia.org/wiki/Z_function ?

No that is not related to. Here are some additional links about the algorithm.

https://contest.cs.cmu.edu/295/tutorials/z-string-matching.pdf
https://codeforces.com/blog/entry/3107?
https://www.hackerearth.com/practice/algorithms/string-algorithm/z-algorithm/tutorial/#:~:text=One%20application%20of%20the%20Z,a%20character%20that%20matches%20nothing.

It is really often used efficient algorithm in competitive programming.

**Comment by nikalosa** [comment]
Because of new merged file (./conversions/decimal_to_any.py), checks have failed for my code.

The log:
./conversions/decimal_to_any.py:104:17: E123 closing bracket does not match indentation of opening bracket's line

**Pull request merged**